src/devices/vdml_imu.c

```diff
@@ -156,3 +184,135 @@ imu_status_e_t imu_get_status(uint8_t port) {
 }
+
+//Reset Functions:
+int32_t imu_reset(uint8_t port){
+    if (!claim_port_try(port - 1, E_DEVICE_IMU)) {
+		return PROS_ERR;
+	}
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	imu_set_heading_offset(port - 1, -vexDeviceImuDegreesGet(device->device_info));
+	imu_set_rotation_offset(port - 1, -vexDeviceImuHeadingGet(device->device_info));
+	imu_set_pitch_offset(port - 1, -imu_get_euler_bl(port).pitch);
+	imu_set_roll_offset(port - 1, -imu_get_euler_bl(port).roll);
+	imu_set_yaw_offset(port - 1, -imu_get_euler_bl(port).yaw);
+	return_port(port - 1, 1);
+}
+
+int32_t imu_reset_heading(uint8_t port){
+    return imu_set_heading(port, 0);
+}
+
+int32_t imu_reset_rotation(uint8_t port){
+    return imu_set_rotation(port, 0);
+}
+
+int32_t imu_reset_pitch(uint8_t port){
+    return imu_set_pitch(port, 0);
+}
+
+int32_t imu_reset_roll(uint8_t port){
+    return imu_set_roll(port, 0);
+}
+
+int32_t imu_reset_yaw(uint8_t port){
+    return imu_set_yaw(port, 0);
+}
+
+//Setter Functions:
+int32_t imu_set_rotation(uint8_t port, double target){
+    if (!claim_port_try(port - 1, E_DEVICE_IMU)) {
+		return PROS_ERR;
+	}
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	ERROR_IMU_STILL_CALIBRATING(port, device, PROS_ERR);
+	imu_set_rotation_offset(port - 1, target - vexDeviceImuHeadingGet(device->device_info));
+	return_port(port - 1, 1);
+}
+
+int32_t imu_set_heading(uint8_t port, double target){
+    if (!claim_port_try(port - 1, E_DEVICE_IMU)) {
+		return PROS_ERR;
+	}
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	ERROR_IMU_STILL_CALIBRATING(port, device, PROS_ERR);
+	imu_set_heading_offset(port - 1, target - vexDeviceImuDegreesGet(device->device_info));
+	return_port(port - 1, 1);
+}
+
+int32_t imu_set_pitch(uint8_t port, double target){
+    if (!claim_port_try(port - 1, E_DEVICE_IMU)) {
+		return PROS_ERR;
+	}
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	ERROR_IMU_STILL_CALIBRATING(port, device, PROS_ERR);
+	imu_set_pitch_offset(port - 1, target - imu_get_euler_bl(port).pitch);
+	return_port(port - 1, 1);
+}
+
+int32_t imu_set_roll(uint8_t port, double target){
+    if (!claim_port_try(port - 1, E_DEVICE_IMU)) {
+		return PROS_ERR;
+	}
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	ERROR_IMU_STILL_CALIBRATING(port, device, PROS_ERR);
+	imu_set_roll_offset(port - 1, target - imu_get_euler_bl(port).roll);
+	return_port(port - 1, 1);
+}
+
+int32_t imu_set_yaw(uint8_t port, double target){
+    if (!claim_port_try(port - 1, E_DEVICE_IMU)) {
+		return PROS_ERR;
+	}
+	v5_smart_device_s_t* device = registry_get_device(port - 1);
+	ERROR_IMU_STILL_CALIBRATING(port, device, PROS_ERR);
+	imu_set_yaw_offset(port - 1, target - imu_get_euler_bl(port).yaw);
+	return_port(port - 1, 1);
+}
+
+//Internal functions for getting IMU offsets (Not thread safe)
+double imu_get_heading_offset(uint8_t port){
+    return ((imu_data_s_t*)registry_get_device(port)->pad)->heading_offset;
```

Comment:
My advice would be to factor getting the imu_data_s_t into a helper, and add an assert that's the struct is smaller then however much memory you get.